explorer/interpreter/type_checker.cpp

```diff
@@ -662,6 +663,21 @@ auto TypeChecker::ArgumentDeduction(
       if (std::find(bindings_to_deduce.begin(), bindings_to_deduce.end(),
                     &var_type.binding()) != bindings_to_deduce.end()) {
         auto [it, success] = deduced.insert({&var_type.binding(), arg});
+
+        const Value* substituted_param_type = Substitute(deduced, param);
+        const auto& binding = cast<VariableType>(*param).binding();
+        if (binding.has_static_type()) {
+          if (!IsTypeOfType(&binding.static_type())) {
+            if (!IsImplicitlyConvertible(substituted_param_type,
+                                         &binding.static_type(), impl_scope,
+                                         false)) {
+              return CompilationError(source_loc)
+                     << "Could not convert argument " << *substituted_param_type
+                     << " to parameter type " << binding.static_type();
```

Comment:
```suggestion
                     << "cannot convert deduced value " << *arg
                     << " for " << binding.name()
                     << " to parameter type " << *substituted_param_type;
```
It'd be nice to list the parameter name here too, since we have it.